cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -459,16 +508,181 @@ class self_comparator {
+   * @return `true` if row from the `lhs` table compares less than row in the `rhs` table
+   */
+  __device__ constexpr weak_ordering operator()(lhs_index_type const lhs_index,
+                                                rhs_index_type const rhs_index) const noexcept
+  {
+    return comp(static_cast<cudf::size_type>(lhs_index), static_cast<cudf::size_type>(rhs_index));
+  }
+
+  /**
+   * @brief Checks whether the row at `rhs_index` in the `rhs` table compares
+   * lexicographically less than the row at `lhs_index` in the `lhs` table.
+   *
+   * @param rhs_index The index of the row in the `rhs` table to examine
+   * @param lhs_index The index of the row in the `lhs` table to examine
+   * @return `true` if row from the `rhs` table compares less than row in the `lhs` table
+   */
+  __device__ constexpr weak_ordering operator()(rhs_index_type const rhs_index,
+                                                lhs_index_type const lhs_index) const noexcept
+  {
+    auto const left_right_ordering =
+      comp(static_cast<cudf::size_type>(lhs_index), static_cast<cudf::size_type>(rhs_index));
+
+    // Invert less/greater values to reflect right to left ordering
+    if (left_right_ordering == weak_ordering::LESS) {
+      return weak_ordering::GREATER;
+    } else if (left_right_ordering == weak_ordering::GREATER) {
+      return weak_ordering::LESS;
+    }
+    return weak_ordering::EQUIVALENT;
+  }
+
+ private:
+  /**
+   * @brief Construct a function object for performing a lexicographic
+   * comparison between the rows of two tables with strongly typed table index
+   * types.
+   *
+   * @param check_nulls Indicates if either input table contains columns with nulls.
+   * @param lhs The first table
+   * @param rhs The second table (may be the same table as `lhs`)
+   * @param depth Optional, device array the same length as a row that contains starting depths of
+   * columns if they're nested, and 0 otherwise.
+   * @param column_order Optional, device array the same length as a row that indicates the desired
+   * ascending/descending order of each column in a row. If `nullopt`, it is assumed all columns are
+   * sorted in ascending order.
+   * @param null_precedence Optional, device array the same length as a row and indicates how null
+   * values compare to all other for every column. If `nullopt`, then null precedence would be
+   * `null_order::BEFORE` for all columns.
+   */
+  two_table_device_row_comparator_adapter(
+    Nullate check_nulls,
+    table_device_view lhs,
+    table_device_view rhs,
+    std::optional<device_span<int const>> depth                  = std::nullopt,
+    std::optional<device_span<order const>> column_order         = std::nullopt,
+    std::optional<device_span<null_order const>> null_precedence = std::nullopt)
+    : comp{check_nulls, lhs, rhs, depth, column_order, null_precedence}
+  {
+  }
+
+  device_row_comparator<Nullate> comp;
```

Comment:
I thought I left this feedback before, but the "adapter" could be simplified by making the type of `comp` a template parameter. This way it's a true `adapter` and it's sole job is to just convert strong type indices appropriately for the underlying comparator. 

```
template <typename C>
concept row_comparator = requires(C c, size_type i, size_type j){
   {c(i,j)} -> std::same_as<weak_ordering>;
};

template <row_comparator C>
struct strong_index_comparator_adapter{
 public:
  __device__ constexpr weak_ordering operator()(lhs_index_type const lhs_index,
                                                rhs_index_type const rhs_index) const noexcept
  {
    return comp(static_cast<cudf::size_type>(lhs_index), static_cast<cudf::size_type>(rhs_index));
  }

  __device__ constexpr weak_ordering operator()(rhs_index_type const rhs_index,
                                                lhs_index_type const lhs_index) const noexcept
  {
    auto const left_right_ordering =
      comp(static_cast<cudf::size_type>(lhs_index), static_cast<cudf::size_type>(rhs_index));

    // Invert less/greater values to reflect right to left ordering
    if (left_right_ordering == weak_ordering::LESS) {
      return weak_ordering::GREATER;
    } else if (left_right_ordering == weak_ordering::GREATER) {
      return weak_ordering::LESS;
    }
    return weak_ordering::EQUIVALENT;
  }

private:
   C const comp
};

Reply:
Ah, nice. This makes the strong type layer and weak ordering layer look and act much more similarly. This is implemented now.